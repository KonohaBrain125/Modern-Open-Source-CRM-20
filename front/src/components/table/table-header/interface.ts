import { ReactNode } from 'react';

export type SortType<SortKey = string> = {
  label: string;
  key: SortKey;
  icon?: ReactNode;
};

export type FilterType<FilterKey = string> = {
  label: string;
  key: FilterKey;
  icon: ReactNode;
};

export type SelectedFilterType = {
  id: string;
  label: string;
  value: string;
  operand: { id: string; label: string };
  icon: ReactNode;
};

export type SelectedSortType<SortField = string> = SortType<SortField> & {
  order: 'asc' | 'desc';
};
