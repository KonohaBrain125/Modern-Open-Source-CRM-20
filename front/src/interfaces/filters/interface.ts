import { ReactNode } from 'react';
import { SearchConfigType, SearchableType } from '../search/interface';
import { Person } from '../entities/person.interface';
import { Company } from '../entities/company.interface';
import { User } from '../entities/user.interface';
import { AnyEntity, BoolExpType } from '../entities/generic.interface';

export type FilterableFieldsType = Person | Company;
export type FilterWhereType = Person | Company | User | AnyEntity;

export type FilterConfigType<
  FilteredType extends FilterableFieldsType,
  WhereType extends FilterWhereType = any,
> = {
  key: string;
  label: string;
  icon: ReactNode;
  operands: FilterOperandType<FilteredType, WhereType>[];
  searchConfig: WhereType extends SearchableType
    ? SearchConfigType<WhereType>
    : null;
  selectedValueRender: (selected: WhereType) => string;
};

export type FilterOperandType<
  FilteredType extends FilterableFieldsType,
  WhereType extends FilterWhereType = AnyEntity,
> =
  | FilterOperandExactMatchType<FilteredType, WhereType>
  | FilterOperandComparativeType<FilteredType, WhereType>;

type FilterOperandExactMatchType<
  FilteredType extends FilterableFieldsType,
  WhereType extends FilterWhereType,
> = {
  label: 'Equal' | 'Not equal';
  id: 'equal' | 'not-equal';
  whereTemplate: (value: WhereType) => BoolExpType<FilteredType>;
};

type FilterOperandComparativeType<
  FilteredType extends FilterableFieldsType,
  WhereType extends FilterWhereType,
> = {
  label: 'Like' | 'Not like' | 'Include';
  id: 'like' | 'not_like' | 'include';
  whereTemplate: (value: WhereType) => BoolExpType<FilteredType>;
};

export type SelectedFilterType<
  FilteredType extends FilterableFieldsType,
  WhereType extends FilterWhereType = AnyEntity,
> = {
  key: string;
  value: WhereType;
  displayValue: string;
  label: string;
  icon: ReactNode;
  operand: FilterOperandType<FilteredType, WhereType>;
};
