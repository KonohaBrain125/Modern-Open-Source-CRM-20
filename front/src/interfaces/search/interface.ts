import { DocumentNode } from 'graphql';
import { ReactNode } from 'react';
import {
  Companies_Bool_Exp,
  People_Bool_Exp,
  Users_Bool_Exp,
} from '../../generated/graphql';
import { Person } from '../entities/person.interface';
import { Company } from '../entities/company.interface';
import { User } from '../entities/user.interface';
import { GqlType } from '../entities/generic.interface';

export type SearchableType = Person | Company | User;

export type SearchConfigType<SearchType extends SearchableType> = {
  query: DocumentNode;
  template: (
    searchInput: string,
  ) => People_Bool_Exp | Companies_Bool_Exp | Users_Bool_Exp;
  resultMapper: (data: GqlType<SearchType>) => {
    value: SearchType;
    render: (value: SearchType) => ReactNode;
  };
};
