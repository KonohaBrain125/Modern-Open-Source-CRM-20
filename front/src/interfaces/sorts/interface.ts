import { ReactNode } from 'react';
import { SortOrder as Order_By } from '../../generated/graphql';

export type SortType<OrderByTemplate> =
  | {
      _type: 'default_sort';
      label: string;
      key: keyof OrderByTemplate & string;
      icon?: ReactNode;
    }
  | {
      _type: 'custom_sort';
      label: string;
      key: string;
      icon?: ReactNode;
      orderByTemplates: Array<(order: Order_By) => OrderByTemplate>;
    };

export type SelectedSortType<OrderByTemplate> = SortType<OrderByTemplate> & {
  order: 'asc' | 'desc';
};
