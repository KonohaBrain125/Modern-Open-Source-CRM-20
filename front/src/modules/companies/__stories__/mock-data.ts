import { Pipeline } from '~/generated/graphql';

export const pipeline = {
  id: 'pipeline-1',
  name: 'pipeline-1',
  pipelineStages: [
    {
      id: 'pipeline-stage-1',
      name: 'New',
      position: 0,
      color: 'red',
    },
    {
      id: 'pipeline-stage-2',
      name: 'Screening',
      position: 1,
      color: 'purple',
    },
    {
      id: 'pipeline-stage-3',
      name: 'Meeting',
      position: 2,
      color: 'sky',
    },
    {
      id: 'pipeline-stage-4',
      name: 'Proposal',
      position: 3,
      color: 'turquoise',
    },
    {
      id: 'pipeline-stage-5',
      name: 'Customer',
      position: 4,
      color: 'yellow',
    },
  ],
} as Pipeline;
