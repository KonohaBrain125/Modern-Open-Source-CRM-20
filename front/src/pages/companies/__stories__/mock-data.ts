import { GraphqlQueryCompany } from '../../../interfaces/company.interface';

export const defaultData: Array<GraphqlQueryCompany> = [
  {
    id: 'f121ab32-fac4-4b8c-9a3d-150c877319c2',
    name: 'ACME',
    domain_name: 'example.com',
    account_owner: {
      id: '91510aa5-ede6-451f-8029-a7fa69e4bad6',
      email: 'john@example.com',
      displayName: 'John Doe',
      __typename: 'User',
    },
    employees: 10,
    address: '1 Infinity Loop, 95014 Cupertino, California',
    created_at: new Date().toUTCString(),
  },
];
