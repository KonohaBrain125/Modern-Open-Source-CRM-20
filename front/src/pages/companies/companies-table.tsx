import { createColumnHelper } from '@tanstack/react-table';
import {
  Company,
  GraphqlQueryCompany,
} from '../../interfaces/company.interface';
import { updateCompany } from '../../services/companies';
import ColumnHead from '../../components/table/ColumnHead';
import Checkbox from '../../components/form/Checkbox';
import CompanyChip from '../../components/chips/CompanyChip';
import EditableText from '../../components/table/editable-cell/EditableText';
import PipeChip from '../../components/chips/PipeChip';
import {
  FaRegBuilding,
  FaCalendar,
  FaLink,
  FaMapPin,
  FaStream,
  FaRegUser,
  FaUsers,
  FaBuilding,
} from 'react-icons/fa';
import ClickableCell from '../../components/table/ClickableCell';
import PersonChip from '../../components/chips/PersonChip';
import EditableChip from '../../components/table/editable-cell/EditableChip';
import {
  FilterType,
  SortType,
} from '../../components/table/table-header/interface';
import {
  Companies_Bool_Exp,
  Companies_Order_By,
} from '../../generated/graphql';
import { SEARCH_COMPANY_QUERY } from '../../services/search/search';

export const availableSorts = [
  {
    key: 'name',
    label: 'Name',
    icon: <FaBuilding />,
    _type: 'default_sort',
  },
  {
    key: 'employees',
    label: 'Employees',
    icon: <FaUsers />,
    _type: 'default_sort',
  },
  {
    key: 'domain_name',
    label: 'Url',
    icon: <FaLink />,
    _type: 'default_sort',
  },
  {
    key: 'address',
    label: 'Address',
    icon: <FaMapPin />,
    _type: 'default_sort',
  },
  {
    key: 'created_at',
    label: 'Creation',
    icon: <FaCalendar />,
    _type: 'default_sort',
  },
] satisfies Array<SortType<Companies_Order_By>>;

export const availableFilters = [
  {
    key: 'company_name',
    label: 'Company',
    icon: <FaBuilding />,
    whereTemplate: (operand, { companyName }) => {
      if (operand.keyWord === 'equal') {
        return {
          name: { _eq: companyName },
        };
      }

      if (operand.keyWord === 'not_equal') {
        return {
          _not: { name: { _eq: companyName } },
        };
      }
    },
    searchQuery: SEARCH_COMPANY_QUERY,
    searchTemplate: (searchInput: string) => ({
      name: { _ilike: `%${searchInput}%` },
    }),
    searchResultMapper: (company: GraphqlQueryCompany) => ({
      displayValue: company.name,
      value: { companyName: company.name },
    }),
    operands: [
      { label: 'Equal', id: 'equal', keyWord: 'equal' },
      { label: 'Not equal', id: 'not-equal', keyWord: 'not_equal' },
    ],
  },
  {
    key: 'domainName',
    label: 'Url',
    icon: <FaLink />,
    whereTemplate: (operand, { domainName }) => {
      if (operand.keyWord === 'equal') {
        return {
          domain_name: { _eq: domainName },
        };
      }

      if (operand.keyWord === 'not_equal') {
        return {
          _not: { domain_name: { _eq: domainName } },
        };
      }
    },
    searchQuery: SEARCH_COMPANY_QUERY,
    searchTemplate: (searchInput: string) => ({
      domain_name: { _ilike: `%${searchInput}%` },
    }),
    searchResultMapper: (company: GraphqlQueryCompany) => ({
      displayValue: company.domain_name,
      value: { domainName: company.domain_name },
    }),
    operands: [
      { label: 'Equal', id: 'equal', keyWord: 'equal' },
      { label: 'Not equal', id: 'not-equal', keyWord: 'not_equal' },
    ],
  },
] satisfies Array<FilterType<Companies_Bool_Exp>>;

const columnHelper = createColumnHelper<Company>();
export const companiesColumns = [
  columnHelper.accessor('id', {
    header: () => (
      <Checkbox id="company-select-all" name="company-select-all" />
    ),
    cell: (props) => (
      <Checkbox
        id={`company-selected-${props.row.original.id}`}
        name={`company-selected-${props.row.original.id}`}
      />
    ),
  }),
  columnHelper.accessor('name', {
    header: () => <ColumnHead viewName="Name" viewIcon={<FaRegBuilding />} />,
    cell: (props) => (
      <EditableChip
        value={props.row.original.name}
        placeholder="Name"
        picture={`https://www.google.com/s2/favicons?domain=${props.row.original.domain_name}&sz=256`}
        changeHandler={(value: string) => {
          const company = props.row.original;
          company.name = value;
          updateCompany(company).catch((error) => console.error(error));
        }}
        ChipComponent={CompanyChip}
      />
    ),
  }),
  columnHelper.accessor('employees', {
    header: () => <ColumnHead viewName="Employees" viewIcon={<FaUsers />} />,
    cell: (props) => (
      <EditableText
        content={props.row.original.employees.toFixed(0)}
        changeHandler={(value) => {
          const company = props.row.original;
          company.employees = parseInt(value);
          updateCompany(company).catch((error) => console.error(error));
        }}
      />
    ),
  }),
  columnHelper.accessor('domain_name', {
    header: () => <ColumnHead viewName="URL" viewIcon={<FaLink />} />,
    cell: (props) => (
      <EditableText
        content={props.row.original.domain_name}
        changeHandler={(value) => {
          const company = props.row.original;
          company.domain_name = value;
          updateCompany(company).catch((error) => console.error(error));
        }}
      />
    ),
  }),
  columnHelper.accessor('address', {
    header: () => <ColumnHead viewName="Address" viewIcon={<FaMapPin />} />,
    cell: (props) => (
      <EditableText
        content={props.row.original.address}
        changeHandler={(value) => {
          const company = props.row.original;
          company.address = value;
          updateCompany(company).catch((error) => console.error(error));
        }}
      />
    ),
  }),
  columnHelper.accessor('opportunities', {
    header: () => (
      <ColumnHead viewName="Opportunities" viewIcon={<FaStream />} />
    ),
    cell: (props) => (
      <ClickableCell href="#">
        {props.row.original.opportunities.map((opportunity) => (
          <PipeChip opportunity={opportunity} />
        ))}
      </ClickableCell>
    ),
  }),
  columnHelper.accessor('creationDate', {
    header: () => <ColumnHead viewName="Creation" viewIcon={<FaCalendar />} />,
    cell: (props) => (
      <ClickableCell href="#">
        {new Intl.DateTimeFormat(undefined, {
          month: 'short',
          day: 'numeric',
          year: 'numeric',
        }).format(props.row.original.creationDate)}
      </ClickableCell>
    ),
  }),
  columnHelper.accessor('accountOwner', {
    header: () => (
      <ColumnHead viewName="Account Owner" viewIcon={<FaRegUser />} />
    ),
    cell: (props) => (
      <ClickableCell href="#">
        <>
          {props.row.original.accountOwner && (
            <PersonChip name={props.row.original.accountOwner?.displayName} />
          )}
        </>
      </ClickableCell>
    ),
  }),
];
