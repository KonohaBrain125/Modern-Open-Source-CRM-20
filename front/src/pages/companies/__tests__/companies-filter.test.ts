import { FilterType } from '../../../components/table/table-header/interface';
import { Companies_Bool_Exp } from '../../../generated/graphql';
import { GraphqlQueryCompany } from '../../../interfaces/company.interface';
import { GraphqlQueryPerson } from '../../../interfaces/person.interface';
import {
  SEARCH_COMPANY_QUERY,
  SEARCH_PEOPLE_QUERY,
} from '../../../services/search/search';
import { mockData } from './__data__/mock-data';
import { availableFilters } from '../companies-table';

function assertFilterUseCompanySearch<FilterValue>(
  filter: FilterType<Companies_Bool_Exp>,
): filter is FilterType<Companies_Bool_Exp> & {
  searchResultMapper: (data: GraphqlQueryCompany) => {
    displayValue: string;
    value: FilterValue;
  };
} {
  return filter.searchQuery === SEARCH_COMPANY_QUERY;
}

function assertFilterUsePeopleSearch<FilterValue>(
  filter: FilterType<Companies_Bool_Exp>,
): filter is FilterType<Companies_Bool_Exp> & {
  searchResultMapper: (data: GraphqlQueryPerson) => {
    displayValue: string;
    value: FilterValue;
  };
} {
  return filter.searchQuery === SEARCH_PEOPLE_QUERY;
}

const AirbnbCompany = mockData.find(
  (user) => user.name === 'Airbnb',
) as GraphqlQueryCompany;

describe('CompaniesFilter', () => {
  for (const filter of availableFilters) {
    it(`should render the filter ${filter.key}`, () => {
      if (assertFilterUseCompanySearch(filter)) {
        const filterSelectedValue = filter.searchResultMapper(mockData[0]);
        for (const operand of filter.operands) {
          expect(
            filter.whereTemplate(operand, filterSelectedValue.value),
          ).toMatchSnapshot();
        }
      }
      if (assertFilterUsePeopleSearch(filter)) {
        const filterSelectedValue = filter.searchResultMapper(AirbnbCompany);
        for (const operand of filter.operands) {
          expect(
            filter.whereTemplate(operand, filterSelectedValue.value),
          ).toMatchSnapshot();
        }
      }
    });
    it(`should render the serch ${filter.key} with the searchValue`, () => {
      expect(filter.searchTemplate('Search value')).toMatchSnapshot();
    });
  }
});
