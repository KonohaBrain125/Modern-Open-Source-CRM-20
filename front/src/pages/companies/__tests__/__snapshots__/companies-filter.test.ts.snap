// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompaniesFilter should render the filter company_name 1`] = `
Object {
  "name": Object {
    "_eq": "Airbnb",
  },
}
`;

exports[`CompaniesFilter should render the filter company_name 2`] = `
Object {
  "_not": Object {
    "name": Object {
      "_eq": "Airbnb",
    },
  },
}
`;

exports[`CompaniesFilter should render the filter domainName 1`] = `
Object {
  "domain_name": Object {
    "_eq": "airbnb.com",
  },
}
`;

exports[`CompaniesFilter should render the filter domainName 2`] = `
Object {
  "_not": Object {
    "domain_name": Object {
      "_eq": "airbnb.com",
    },
  },
}
`;

exports[`CompaniesFilter should render the serch company_name with the searchValue 1`] = `
Object {
  "name": Object {
    "_ilike": "%Search value%",
  },
}
`;

exports[`CompaniesFilter should render the serch domainName with the searchValue 1`] = `
Object {
  "domain_name": Object {
    "_ilike": "%Search value%",
  },
}
`;
